File: VietOCR.NET/tags/Release-2.0/GUIWithSpellcheck.cs
/**
 * Copyright @ 2008 Quan Nguyen
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using Net.SourceForge.Vietpad.Utilities;

namespace VietOCR.NET
{
    public partial class GUIWithSpellcheck : VietOCR.NET.GUIWithSettings
    {
        private int start, end;
        private SpellChecker sp;
        private string curWord;

        public GUIWithSpellcheck()
        {
            InitializeComponent();
        }

        protected override void contextMenuStrip1_Opening(object sender, CancelEventArgs e)
        {
            try
            {
                this.contextMenuStrip1.Items.Clear();
                if (this.toolStripButtonSpellCheck.Checked)
                {
                    int offset = this.textBox1.GetCharIndexFromPosition(pointClicked);
                    BreakIterator boundary = BreakIterator.GetWordInstance();
                    boundary.Text = this.textBox1.Text;
                    end = boundary.Following(offset);

                    if (end != BreakIterator.DONE)
                    {
                        start = boundary.Previous();
                    }

                    curWord = this.textBox1.Text.Substring(start, end - start);
                    makeSuggestions(curWord);
                }
            }
            finally
            {
                // load standard menu items
                this.contextMenuStrip1.RepopulateContextMenu();
            }
        }

        /// <summary>
        /// Populates suggestions at top of context menu.
        /// </summary>
        /// <param name="curWord"></param>
        void makeSuggestions(string curWord)
        {
            if (sp == null || curWord == null || curWord.Trim().Length == 0)
            {
                return;
            }

            List<String> suggests = sp.Suggest(curWord);
            if (suggests == null || suggests.Count == 0)
            {
                return;
            }

            foreach (string word in suggests)
            {
                ToolStripMenuItem item = new ToolStripMenuItem(word);
                item.Font = new Font(item.Font, FontStyle.Bold);
                item.Click += new EventHandler(item_Click);
                this.contextMenuStrip1.Items.Add(item);
            }
            this.contextMenuStrip1.Items.Add("-");

            ToolStripMenuItem item1 = new ToolStripMenuItem(Properties.Resources.Ignore_All);
            item1.Tag = "ignore.word";
            item1.Click += new EventHandler(item_Click);
            this.contextMenuStrip1.Items.Add(item1);

            item1 = new ToolStripMenuItem(Properties.Resources.Add_to_Dictionary);
            item1.Tag = "add.word";
            item1.Click += new EventHandler(item_Click);
            this.contextMenuStrip1.Items.Add(item1);
            this.contextMenuStrip1.Items.Add("-");
        }

        void item_Click(object sender, EventArgs e)
        {
            ToolStripItem item = (ToolStripItem)sender;
            object command = item.Tag;

            if (command == null)
            {
                this.textBox1.Select(start, end - start);
                this.textBox1.SelectedText = item.Text;
            }
            else if (command.ToString() == "ignore.word")
            {
                sp.IgnoreWord(curWord);
            }
            else if (command.ToString() == "add.word")
            {
                sp.AddWord(curWord);
            }

            sp.SpellCheck();
        }

        protected override void toolStripButtonSpellCheck_Click(object sender, EventArgs e)
        {
            string localeId = null;

            if (LookupISO_3_1_Codes.ContainsKey(curLangCode))
            {
                localeId = LookupISO_3_1_Codes[curLangCode];
            }
            else if (LookupISO_3_1_Codes.ContainsKey(curLangCode.Substring(0, 3)))
            {
                localeId = LookupISO_3_1_Codes[curLangCode.Substring(0, 3)];
            }

            if (localeId == null)
            {
                MessageBox.Show("Need to add an entry in Data/ISO639-1.xml file.");
                return;
            }

            sp = new SpellChecker(this.textBox1, localeId);

            if (this.toolStripButtonSpellCheck.Checked)
            {
                sp.EnableSpellCheck();
            }
            else
            {
                sp.DisableSpellCheck();
            }
            this.textBox1.Refresh();
        }
    }
}

File: VietOCR/tags/Release-3.1/src/net/sourceforge/vietocr/GuiWithSpellcheck.java
/**
 * Copyright @ 2008 Quan Nguyen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.sourceforge.vietocr;

import java.awt.Font;
import java.awt.Point;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.List;
import java.util.Properties;
import javax.swing.JMenuItem;
import javax.swing.JOptionPane;
import javax.swing.text.BadLocationException;

public class GuiWithSpellcheck extends GuiWithSettings {

    private int start, end;
    private SpellChecker sp;

    @Override
    void populatePopupMenuWithSuggestions(Point pointClicked) {
        try {
            popup.removeAll();
            if (this.jToggleButtonSpellCheck.isSelected()) {
                int offset = jTextArea1.viewToModel(pointClicked);
                start = javax.swing.text.Utilities.getWordStart(jTextArea1, offset);
                end = javax.swing.text.Utilities.getWordEnd(jTextArea1, offset);
                String curWord = jTextArea1.getDocument().getText(start, end - start);
                makeSuggestions(curWord);
            }
        } catch (BadLocationException e) {
        } finally {
            // load standard menu items
            repopulatePopupMenu();
        }
    }

    /**
     * Populates suggestions at top of context menu.
     * @param curWord
     */
    void makeSuggestions(final String curWord) {
        if (sp == null || curWord == null || curWord.trim().length() == 0) {
            return;
        }

        List<String> suggests = sp.suggest(curWord);
        if (suggests == null || suggests.isEmpty()) {
            return;
        }

        ActionListener correctLst = new ActionListener() {

            @Override
            public void actionPerformed(ActionEvent ae) {
                String selectedWord = ae.getActionCommand();
                if (selectedWord.equals("ignore.word")) {
                    sp.ignoreWord(curWord);
                } else if (selectedWord.equals("add.word")) {
                    sp.addWord(curWord);
                } else {
                    jTextArea1.select(start, end);
                    jTextArea1.replaceSelection(selectedWord);
                }
                sp.spellCheck();
            }
        };

        for (String word : suggests) {
            JMenuItem item = new JMenuItem(word);
            item.setFont(item.getFont().deriveFont(Font.BOLD));
            item.addActionListener(correctLst);
            popup.add(item);
        }

        popup.addSeparator();
        JMenuItem item = new JMenuItem(bundle.getString("Ignore_All"));
        item.setActionCommand("ignore.word");
        item.addActionListener(correctLst);
        popup.add(item);
        item = new JMenuItem(bundle.getString("Add_to_Dictionary"));
        item.setActionCommand("add.word");
        item.addActionListener(correctLst);
        popup.add(item);
        popup.addSeparator();
    }

    @Override
    void spellCheckActionPerformed() {
        Properties lookupISO_3_1_Codes = getLookupISO_3_1_Codes();
        String localeId = null;

        if (lookupISO_3_1_Codes.containsKey(curLangCode)) {
            localeId = lookupISO_3_1_Codes.getProperty(curLangCode);
        } else if (lookupISO_3_1_Codes.containsKey(curLangCode.substring(0, 3))) {
            localeId = lookupISO_3_1_Codes.getProperty(curLangCode.substring(0, 3));
        }
        if (localeId == null) {
            JOptionPane.showMessageDialog(null, "Need to add an entry in data/ISO639-1.xml file.", Gui.APP_NAME, JOptionPane.ERROR_MESSAGE);
            return;
        }
        
        sp = new SpellChecker(this.jTextArea1, localeId);
        if (this.jToggleButtonSpellCheck.isSelected()) {
            sp.enableSpellCheck();
        } else {
            sp.disableSpellCheck();
        }
        this.jTextArea1.repaint();
    }
}
